core/src/main/java/com/linecorp/armeria/internal/common/AbortedHttpResponse.java

```diff
@@ -15,12 +15,30 @@
  */
 package com.linecorp.armeria.internal.common;
 
+import java.util.function.Function;
+
 import com.linecorp.armeria.common.HttpObject;
 import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.ResponseHeaders;
 import com.linecorp.armeria.internal.common.stream.AbortedStreamMessage;
+import com.linecorp.armeria.server.HttpResponseException;
 
 public final class AbortedHttpResponse extends AbortedStreamMessage<HttpObject> implements HttpResponse {
     public AbortedHttpResponse(Throwable cause) {
         super(cause);
     }
+
+    @Override
+    public HttpResponse mapHeaders(Function<? super ResponseHeaders, ? extends ResponseHeaders> function) {
```

Comment:
I believe this method is superfluous, but I'm not sure. It was added after a very short discussion with @ikhoon, but I believe the case is already handled by `HttpResponse.mapHeaders()` (if not, I need better tests, because they are not affected by the existence of this)

Reply:
Could you also override `mapData`, `mapTrailers`, and other operators possible?

Reply:
Addtionally, I think we can directly covert an `HttpResponseException` whose cause is null to `HttpResponse` in `HttpResponse ofFailure(Throwable cause)`
```java
HttpResponse ofFailure(Throwable cause) {
    if (cause instanceof HttpResponseException) {
        final HttpResponseException httpResponseException = (HttpResponseException) getCause();
        if (httpResponseException.getCause() == null) {
            return httpResponseException.httpResponse();
        }
    }
    ...
}
```

Reply:
@ikhoon Is this more or less what you imagined?

Is there something else, e.g. `FilteredHttpResponse`, that could require special handling?

Reply:
This is what I imagined. Looks good overall.

Reply:
@ikhoon I'm having serious problems getting this working, especially with a real server, I get `FixedHttpResponse` which is a different implemented to `AbortedHttpResponse`/`HttpResponse`. It leads me to believe there are other more exotic classes being used as well. Maybe it's better to put all logic in HttpResponse?

Reply:
>  I get FixedHttpResponse which is a different implemented than AbortedHttpResponse/HttpResponse.

I don't think an aborted `FixedHttpResponse`. Sorry for the wrong direction. 😅 I was worried about additional overheads on `.mapError`.
Since `FuseableStreamMessage` merges multiple `map*` operations into a `StreamMessage`, good to use `mapHeaders(...).mapError(...)`.

Reply:
We missed overriding `filter`, `peekXXX` operators. It might be tedious to override all operators and easy to miss new operations.
https://github.com/line/armeria/pull/4074/files#r802421272
I am leaning toward https://github.com/line/armeria/pull/4074/files#r802421272 style to avoid overriding those operators.
For example:
```java
HttpResponse ofFailure(Throwable cause) {
    if (cause instanceof HttpResponseException) {
        // Need to fix CapturedServiceException to set it for the only first time.
        CapturedServiceException.maybeSet(reqCtx, cause);
        return httpResponseException.httpResponse();
    } else {
        return new AbortedHttpResponse(cause);
    }
}
```

Reply:
My previous comment was wrong. We may or may not get the reference of `reqCtx`. I think we have to need to override stream operations at the moment.